### ISSUE TITLE
Issue with testing a consuming Ruby module

### ISSUE THREAD

**Issue by thewyzard44** [lxc/lxd#3791]
# Required information

 * Distribution: Ubuntu 16.04
 * The output of "lxc info" or if that fails:
config:
  core.https_address: '[::]:8443'
api_extensions:
- id_map
api_status: stable
api_version: "1.0"
auth: trusted
public: false
environment:
  addresses:
  - 10.0.2.15:8443
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
    <truncated>
    -----END CERTIFICATE-----
  certificate_fingerprint: c9eeb72e37391f39fa70e4942d05e91899903f8691f2f88276acb05781ee0ef5
  driver: lxc
  driver_version: 2.0.8
  kernel: Linux
  kernel_architecture: x86_64
  kernel_version: 4.4.0-93-generic
  server: lxd
  server_pid: 1913
  server_version: 2.0.10
  storage: dir
  storage_version: ""

# Issue description

/issue nexussw/lxd-common#1

# Steps to reproduce

 1. clone git@github.com:nexussw/lxd-common
 2. cd lxd-common
 3. vagrant up
 4. vagrant ssh -c 'cd /vagrant; rake spec'

# Information to attach

refer to above linked issue
tl;dr: if a stop happens too soon after a start, `lxc stop` hangs

**Comment by stgraber** [comment]
"lxc stop" only sends SIGPWR (or similar) to PID1 in the container. If PID1 in the container hasn't yet setup its signal handler, then nothing will happen and "lxc stop" will hang.

This is an unfortunate but expected behavior. POSIX signals are unidirectional so there's no way for LXD to know if PID1 is doing something about it.

I suspect in your case you actually want "lxc stop --force" which instead has LXD kill PID1 in the container, causing the container to immediately stop, regardless of what PID1 or in fact any process may be doing in the container.

**Comment by thewyzard44** [comment]
this context is under the hood of a chef provisioning driver where stop probably wants a graceful shutdown - but also in that context, will not normally happen within 2-3 seconds of the start command.  I do do a force if chef wants to destroy, and that does work well.

Thank you for the insight!  I'll look into kill 1 - sounds brutish, but i likely won't put it into production and only do it in testing unless i can convince myself that sleep 10 is acceptable  xD  Now that I know the root cause i can sleep easier.

I accept your answer and I'll close this after the weekend if no one else chimes in with a good idea

**Comment by thewyzard44** [comment]
i do have to wonder, though, if you should be returning from `start` in some 'sane' state that is ready to reliably respond to additional commands

**Comment by jfgibbins** [comment]
I use the following code to check that cloud-init is finished, before preceeding with any other commands.  You could modify or implement something similiar to verify that the container has actually booted first, before trying to run other commands, including shutdown:

"while [ ! -f /var/lib/cloud/instance/boot-finished ]; do echo 'Waiting for cloud-init...'; sleep 5; done;"

**Comment by jfgibbins** [comment]
fyi, the sleep amount could be anything, mine is at five, because it shows at equal intervals with terraform reporting, and my cloud-init usually runs quite a few things, anywhere from 3 secs up to 5-10 mins, depending on if the container is installing applications

**Comment by stgraber** [comment]
The problem is that there is no way for LXD to know when the container is "ready" as that completely depends on what's running in the container. There's no way for LXD to check whether the container's init system is going to be responding to a particular signal or ignore it.

So your best bet is indeed to do your own check based on what matters to you, or call "lxc stop --timeout=30" (or whatever timeout you want), followed with a call to "lxc stop --force" to kill the container if the init system is unresponsive.

**Comment by stgraber** [comment]
Our behavior here is very similar to what you'd have with a VM. A VM hypervisor will say that the VM is running as soon as it's started executing the virtual BIOS. A VM shutdown just sends an ACPi signal, if that's ignore by the VM, then it never goes down, unless you decide to kill it.

**Comment by thewyzard44** [comment]
excellent parallel!!  So After sleeping on it, I'm leaning towards something like:
(pseudocoded)
```ruby
        def stop_container(container_id, options)
          loop do
            return if container_status(container_id) == 'stopped'
            inner_transport.execute("lxc stop #{container_id} --timeout=10")
            if timeout_exceeded # supplied by options - 60 seconds or so by default
              return inner_transport.execute("lxc stop #{container_id} --force") if options[:force]
              raise 'command timeout'
            end
          end
        end
```
Any issue that you can think of with sending too many SIGPWR's?  They just go into oblivion if they're not received?  Or if one is received and it takes longer than --timeout=10 to shutdown gracefully, will a subsequent SIGPWR cause issues?

RE: using cloud-init...  I also like the sounds of that, but I'm not doing anything to convolute the init process, myself, which doesn't take too long by default.  So I think that's more of a sanity check on the part of my caller.  If 'they' slow it down, they're likely to be waiting for it themselves, and won't be sending a stop.  This would be a route one could take at a higher level than this module.

**Comment by stgraber** [comment]
I think most init systems will just ignore subsequent signals, so it should be fine to do it.

**Comment by thewyzard44** [comment]
thank you!!  I'll use --timeout=1 in my testing in an effort to flush those issues out
